kornia/color/zca.py

```diff
@@ -0,0 +1,243 @@
+
+        This implementation uses :class:`~torch.svd` which yields NaNs in the backwards step
+        if the sigular values are not unique.
+
+    References:
+
+        [1] `Stanford PCA & ZCA whitening tutorial <http://ufldl.stanford.edu/tutorial/unsupervised/PCAWhitening/>`_
+
+    """
+
+    def __init__(self, eps: float = 1e-7, biased: bool = False, detach_transforms: bool = True) -> None:
+
+        super(ZCAWhitening, self).__init__()
+
+        self.eps: float = eps
+        self.biased: bool = biased
+        self.detach_transforms: bool = detach_transforms
+        self.fitted = False
+
+    def fit(self, x: torch.Tensor):  # type: ignore
+        r"""
+
+        Fits ZCA whitening matrices to the data.
+
+        args:
+
+            x (torch.Tensor): Input data
+
+        returns:
+            ZCAWhiten: returns a fitted ZCAWhiten object instance.
+        """
+
+        T, mean, T_inv = zca_mean(x, self.eps, self.biased, compute_inv=True)
+
+        self.mean: torch.Tensor = mean
+        self.transform: torch.Tensor = T
+        self.transform_inv: torch.Tensor = T_inv
+
+        if self.detach_transforms:
+            self.mean = self.mean.detach()
+            self.transform = self.transform.detach()
+            self.transform_inv = self.transform_inv.detach()
+
+        self.fitted = True
+
+        return self
+
+    def forward(self, x: torch.Tensor, include_fit: bool = False) -> torch.Tensor:  # type: ignore
+        r"""
+
+        Applies the whitening transform to the data
+
+        args:
+
+            x (torch.Tensor): Input data
+            include_fit (bool): Indicates whether to fit the data as part of the forward pass
+
+        returns:
+
+            torch.Tensor : The transformed data
+
+        """
+
+        if include_fit:
+            self.fit(x)
+
+        if not self.fitted:
+            raise RuntimeError("Needs to be fitted first before running. Please call fit or set include_fit to True.")
+
+        y = zca_whiten(x, self.mean, self.transform)
+
+        return y
+
+    def inverse_transform(self, x: torch.Tensor) -> torch.Tensor:
+        r"""
+
+        Applies the inverse transform to the whitened data.
+
+        args:
+            x (torch.Tensor): Whitened data
+            include_fit (bool): Indicates whether to fit the data as part of the forward pass
+
+        returns:
+            torch.Tensor: original data
+
+
+
+        """
+
+        if not self.fitted:
+            raise RuntimeError("Needs to be fitted first before running. Please call fit or set include_fit to True.")
+
+        num_features: int = reduce(lambda a, b: a * b, x.size()[1::])
+
+        x_flat: torch.Tensor = torch.reshape(x, (-1, num_features))
+
+        y: torch.Tensor = (x_flat).mm(self.transform_inv) + self.mean
+
+        y = y.reshape(x.size())
+
+        return y
+
+
+def zca_mean(inp: torch.Tensor, eps: float = 1e-7, biased: bool = False,
+             compute_inv: bool = False) -> Tuple[torch.Tensor, ...]:
```

Comment:
Besides, `Tuple[torch.Tensor, ...]:` it's a bit confusing to understand what the function actually returns for someone that doesn't know. Then, we have to check what happen with JIT since the return type is variable (this is problem we might face in the whole library)

Reply:
I'll check out how  ```Tuple[torch.Tensor, ...]```interacts with jit. I agree with the proposed changes for ```zca_mean``` since it would make switching between vanilla zca and the zca seen in decorrelated batchnorm easier. I'm assuming we'll apply similar argument changes to the module as well.

Reply:
After looking into it a bit more, I have some lingering question and clarifications. Assuming the dim arg denotes the sample/batch axis, dim = 0, keepdim=True,and following similar output to torch.std_mean, for an input tensor of shape (2,3,4,5), the mean has a shape (1,3,4,5). Should the zca transformation be returned as a (60,60) matrix and keepdim only affects the shape of the mean?

Reply:
probably yes. In this case I guess the output should be specified by (B, C*H*W). /cc @ducha-aiki any suggestions in this ?